csrc/scheduler/matmul.cpp

```diff
@@ -438,6 +438,104 @@ void scheduleProlog(TensorView* shared_mem_tv, const MatmulParams& params) {
 
 } // namespace
 
+namespace {
+int getKAxis(const MmaLayout& layout) {
+  switch (layout) {
+    //! NT : K,M x K,N -> K,M,N
+    case MmaLayout::NT:
+      return 0;
+    //! TT : M,K X K,N -> M,K,N
+    case MmaLayout::TT:
+      return 1;
+    //! TN : M,K X N,K -> M,N,K
+    case MmaLayout::TN:
+      return 2;
+    default:
+      TORCH_CHECK(false, "unsupported data layout.");
+  }
+}
+void scheduleSplitKReduction(
+    TensorView* tv,
+    const MmaLayout& layout,
+    const MatMulTileOptions& gemm_tile,
+    const MatmulParams& params) {
+  // This function is to schedule the reduction part in split-k
+  // step-1, call merge 3 times to merge last 4 domains
+  // id=  0   1   2   3   4   5   6   7  8  9    10
+  // tv= [Moo Moi Noo Noi SKo Mwo Nwo Mw Nw Mist Nist]
+  // Moo = m / cta_tile_m / grid_dim_m
+  // Moi = grid_dim_m
+  // Noo = n / cta_tile_n / grid_dim_n
+  // Noi = grid_dim_n
+  // SKo = split_k_factor
+  // Mwo = cta_tile_m / warp_tile_m
+  // Nwo = cta_tile_n / warp_tile_n
+  // Mw = warp_tile_m / instruction_tile_m
+  // Nw = warp_tile_n / instruction_tile_n
+  // Mist = instruction_tile_m
+  // Nist = instruction_tile_n
+
+  // reset leaf domain leads to simple index
+  // for (int i = 0; i < 3; i++) {
+  //   tv->merge(7);
+  // }
+  tv->domain()->resetLeafDomain();
```

Comment:
I agree. I think in general a good practice is to find a way to build something correct, instead of building something wrong then reset. If our existing infrastructure is not sufficient to build that correct thing, redesign our infrastructure to make it easy to build that correct thing. This practice might not be the most convenient for a single task, but should be much better for longer term robustness.

Reply:
mark it as TODO. This reset is for perf not for convenient. Without reset, all the split/mege histories will be tracked in the indexing and lead to perf drop. The best way is your suggestion "create this tensor before splitting M and N domains". Another approach I used previously is just merge last four domains, then our reduction schedule is limited since the first several domains are already being transformed.